test/deploy.py

```diff
@@ -0,0 +1,569 @@
+    return "".join(
+        random.choices(string.ascii_uppercase + string.digits, k=length)
+    )
+
+
+class Cleaner:
+    def __init__(self, model_file_path: string) -> None:
+        if os.path.exists(model_file_path):
+            # init client
+            self.client = pylxd.Client()
+            self.clean(model_file_path)
+        else:
+            print("Model File {} does not exists.".format(model_file_path))
+
+    def clean(self, model_file_path: string) -> None:
+        """Clean LXC objects mentioned in the model file."""
+        with open(model_file_path, "r") as model_file:
+            model = json.loads(model_file.read())
+
+        if "vm_name" in model:
+            vm_name = model["vm_name"]
+            if self.client.instances.exists(vm_name):
+                print("Deleting VM {}".format(vm_name))
+                vm = self.client.virtual_machines.get(vm_name)
+                vm.stop(wait=True)
+                vm.delete(wait=True)
+
+        if "profile" in model:
+            profile_name = model["profile"]
+            if self.client.profiles.exists(profile_name):
+                print("Deleting VM Profile {}".format(vm_name))
+                self.client.profiles.get(profile_name).delete()
+
+        if "storage_pool" in model:
+            pool_name = model["storage_pool"]
+            if self.client.storage_pools.exists(pool_name):
+                pool = self.client.storage_pools.get(pool_name)
+
+                if "volumes" in model:
+                    volumes = model["volumes"]
+                    for volume in volumes:
+                        # Delete Volume.
+                        print(
+                            "Deleting Volume {} from Pool {}"
+                            .format(volume, pool_name)
+                        )
+                        pool.volumes.get("custom", volume).delete()
+
+                # Delete Storage pool
+                print("Deleting Pool {}".format(pool_name))
+                pool.delete()
+
+
+class DeployRunner:
+    # LXD Vars
+    deploy_tag = "oci-testing-" + _get_random_string(4)
+    complete_repo_path = "/home/ubuntu"
+    cwd = os.getcwd()
+    usr = pwd.getpwuid(os.getuid())[0]
+    modelFilePath = ""
+    model = dict()
+
+    def __init__(self) -> None:
+        # Check LXD installed on host.
+        self.check_snaps_installed()
+        # Check if current user is part of the lxd user group.
+        self.check_user_in_group()
+        # init client
+        self.client = pylxd.Client()
+        # Check if LXD is initialised.
+        self.check_lxd_initialised()
+        # File to store LXD virtual resource references.
+        self.modelFilePath = "{}/model-{}.json".format(
+            self.cwd, _get_random_string(4)
+        )
+
+    def save_model_json(self):
+        """Save vm resource dictionary to json file."""
+        with open(self.modelFilePath, "w") as model_file:
+            model_file.write(json.dumps(self.model, indent=4))
```

Comment:
Nitpick: You can use `json.dump(self.model, model_file, indent=4)` and avoid the pointless conversion to string.

Reply:
Ack